crates/solidity/outputs/cargo/crate/src/extensions/utils/version/mod.rs

```diff
@@ -0,0 +1,438 @@
+    fn hyphen_range(
+        lower_version: &PartialVersion,
+        upper_version: &PartialVersion,
+    ) -> ComparatorSet {
+        let lower = Comparator {
+            version: lower_version.lower_bound(),
+            op: Operator::GtEq,
+        };
+
+        let upper = if upper_version.is_complete() {
+            Comparator {
+                version: upper_version.lower_bound(),
+                op: Operator::LtEq,
+            }
+        } else {
+            upper_version
+                .tilde_upper_bound()
+                .map_or(Comparator::wild(), |version| Comparator {
+                    version,
+                    op: Operator::Lt,
+                })
+        };
+
+        ComparatorSet::bounds(lower, upper)
+    }
+
+    /// Create a `ComparatorSet` that represents the range described by `[partial]`.
+    /// A partially specified version represents a version range, sometimes called an x-range.
+    /// `1.x == >=1.0.0 <2.0.0`
+    /// `0.4.x == >=0.4.0 <0.5.0`
+    /// `2.1 == >=2.1.0 < 2.2.0`
+    fn partial_range(partial: &PartialVersion) -> ComparatorSet {
+        if partial.major.is_wild() {
+            return ComparatorSet::single(Comparator::default());
+        }
+
+        if partial.is_complete() {
+            return ComparatorSet::single(Comparator {
+                version: partial.lower_bound(),
+                op: Operator::Eq,
+            });
+        }
+
+        let lower = Comparator {
+            version: partial.lower_bound(),
+            op: Operator::GtEq,
+        };
+
+        let upper = Comparator {
+            version: partial.tilde_upper_bound().unwrap(),
+            op: Operator::Lt,
+        };
+
+        ComparatorSet::bounds(lower, upper)
+    }
+
+    fn merge(&mut self, other: &ComparatorSet) {
+        for comp in &other.comparators {
+            self.comparators.push(comp.clone());
+        }
+    }
+
+    fn matches(&self, test_version: &Version) -> bool {
+        if self.comparators.is_empty() {
+            false
+        } else {
+            self.comparators.iter().all(|cmp| cmp.matches(test_version))
+        }
+    }
+}
+
+impl Display for ComparatorSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "[")?;
+        for (i, c) in self.comparators.iter().enumerate() {
+            write!(f, "{c}")?;
+            if i < self.comparators.len() - 1 {
+                write!(f, ", ")?;
+            }
+        }
+        write!(f, "]")
+    }
+}
+
+impl PartialEq for ComparatorSet {
+    fn eq(&self, other: &ComparatorSet) -> bool {
+        for comparator in &self.comparators {
+            if !other.comparators.contains(comparator) {
+                return false;
+            }
+        }
+
+        for comparator in &other.comparators {
+            if !self.comparators.contains(comparator) {
+                return false;
+            }
+        }
+
+        true
```

Comment:
Didn't notice before, but it's more efficient to do:

```suggestion
        if self.comparators.len() != other.comparators.len() {
            return false;
        }
        for comparator in &self.comparators {
            if !other.comparators.contains(comparator) {
                return false;
            }
        }

        true
```